### PR TITLE
Optimizations

### PR DESCRIPTION
**Pull request by pienkowb**
In order to speed up the indexing process, this pull request implements 3 optimizations:

* Getting rid of N+1 queries in the aggregator tasks
* Load balancing requests among multiple nodes
* Assigning a dedicated node to each task

### CODE REVIEW

**Review by pacmessica — COMMENTED** [review]
This seems like a different solution to what we discussed a few weeks ago - I don't see multiple workers/fetchers. What I was expecting was something like separating the fetching data from the node part entirely out of the pipeline so we would be able to fetch data from multiple heights at once, and then the fetcher stage in the pipeline would fetch this prefetched data and the pipeline would run as normal processing the heights sequentially.  Did you also try doing some kind of multi-worker solution?

Code seems fine, but I'm concerned this won't significantly improve the speed- there's such a big lag in celo 😅. Did you measure the speed improvement?

File: indexer/pipeline.go [hunk 1]
```diff
@@ -72,16 +72,16 @@ func NewPipeline(
 
 	// Setup stage
 	p.AddStage(
-		pipeline.NewStageWithTasks(pipeline.StageSetup, pipeline.RetryingTask(NewHeightMetaRetrieverTask(client), isTransient, maxRetries)),
+		pipeline.NewStageWithTasks(pipeline.StageSetup, pipeline.RetryingTask(NewHeightMetaRetrieverTask(client.WithAssignedNode(0)), isTransient, maxRetries)),
```
[the remark this answers is not in the document]
Comment (pienkowb) [reply indexer/pipeline.go:75]: Yeah, it's a way to assign specific nodes to different tasks. In the worst case scenario (if there's only one node), the same node will be used in all tasks.

**Comment by pienkowb** [comment]
> This seems like a different solution to what we discussed a few weeks ago - I don't see multiple workers/fetchers. What I was expecting was something like separating the fetching data from the node part entirely out of the pipeline so we would be able to fetch data from multiple heights at once, and then the fetcher stage in the pipeline would fetch this prefetched data and the pipeline would run as normal processing the heights sequentially. Did you also try doing some kind of multi-worker solution?

These are the initial optimizations. Once we merge the code I wrote for Filecoin to the indexing engine, we can use it here.

> Code seems fine, but I'm concerned this won't significantly improve the speed- there's such a big lag in celo 😅. Did you measure the speed improvement?

With different nodes assigned to different tasks there was 30% improvement in the indexing speed (on my local machine). I agree it's not enough to reduce the lag, we certainly need multiple fetching workers.

**Comment by pacmessica** [comment]
> > This seems like a different solution to what we discussed a few weeks ago - I don't see multiple workers/fetchers. What I was expecting was something like separating the fetching data from the node part entirely out of the pipeline so we would be able to fetch data from multiple heights at once, and then the fetcher stage in the pipeline would fetch this prefetched data and the pipeline would run as normal processing the heights sequentially. Did you also try doing some kind of multi-worker solution?
> 
> These are the initial optimizations. Once we merge the code I wrote for Filecoin to the indexing engine, we can use it here.
> 
> > Code seems fine, but I'm concerned this won't significantly improve the speed- there's such a big lag in celo 😅. Did you measure the speed improvement?
> 
> With different nodes assigned to different tasks there was 30% improvement in the indexing speed (on my local machine). I agree it's not enough to reduce the lag, we certainly need multiple fetching workers.

ah ok, so the multiple workers code will be in indexing engine? Does that not cancel out the multiple node solution you have here?

**Comment by pienkowb** [comment]
> > > This seems like a different solution to what we discussed a few weeks ago - I don't see multiple workers/fetchers. What I was expecting was something like separating the fetching data from the node part entirely out of the pipeline so we would be able to fetch data from multiple heights at once, and then the fetcher stage in the pipeline would fetch this prefetched data and the pipeline would run as normal processing the heights sequentially. Did you also try doing some kind of multi-worker solution?
> > 
> > 
> > These are the initial optimizations. Once we merge the code I wrote for Filecoin to the indexing engine, we can use it here.
> > > Code seems fine, but I'm concerned this won't significantly improve the speed- there's such a big lag in celo 😅. Did you measure the speed improvement?
> > 
> > 
> > With different nodes assigned to different tasks there was 30% improvement in the indexing speed (on my local machine). I agree it's not enough to reduce the lag, we certainly need multiple fetching workers.
> 
> ah ok, so the multiple workers code will be in indexing engine? Does that not cancel out the multiple node solution you have here?

These two solutions complement each other but we might as well remove this once we have the multi-worker setup in place. As for the database optimizations, they are useful in both scenarios.

**Review by pacmessica — APPROVED** [review]

**Pull request merged**